File: apps/lasso_ps/lasso_ps.cpp
// Copyright (c) 2014, Sailing Lab
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright
// notice, this list of conditions and the following disclaimer in the
// documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the <ORGANIZATION> nor the names of its contributors
// may be used to endorse or promote products derived from this software
// without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
/*
 * Coordinate descent-based Lasso application built on Petuum Parameter Server.
 * This is implementation of parallel Lasso in:
 * Ho, Qirong, et al. "More effective distributed ml via a stale synchronous parallel parameter server." Advances in Neural Information Processing Systems. 2013.
 *
 * Author: seunghak
 */
#include <vector>
#include <fstream>

#include <petuum_ps/include/petuum_ps.hpp>
#include <gflags/gflags.h>
#include <glog/logging.h>
#include <pthread.h>

#include <boost/date_time/posix_time/posix_time.hpp>
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/uniform_real.hpp>
#include <boost/random/variate_generator.hpp>

#include "spmat.hpp"

// Command-line flags
DEFINE_string(hostfile, "", "Path to Petuum PS server configuration file");
DEFINE_string(datafile_X, "", "Input X");
DEFINE_string(datafile_Y, "", "Output Y");
DEFINE_string(output_prefix, "", "Output results with this prefix. If the prefix is X, the output will be called X.BETA");
DEFINE_int32(num_rows, 1000, "Number of Rows in X");
DEFINE_int32(num_cols, 1000, "Number of Columns in X");
DEFINE_int32(num_clients, 1, "Total number of clients");
DEFINE_int32(num_worker_threads, 1, "Number of worker threads per client");
DEFINE_int32(client_id, 0, "This client's ID");
DEFINE_double(lambda, 0.01, "Lambda");
DEFINE_int32(num_iterations, 100, "Number of iterations");
DEFINE_int32(staleness, 0, "Staleness");

// Typedefs
typedef std::pair<int,float> int_pair_float_t;

// Data variables
col_spmat Xcol(FLAGS_num_rows,FLAGS_num_cols);
col_spmat Ycol(FLAGS_num_rows,1);
col_spmat residual(FLAGS_num_rows,1);

// Returns the number of workers threads across all clients
int get_total_num_workers() {
  return FLAGS_num_clients * FLAGS_num_worker_threads;
}
// Returns the global thread ID of this worker
int get_global_worker_id(int local_thread_id) {
  return (FLAGS_client_id * FLAGS_num_worker_threads) + local_thread_id;
}

void standardization_X() {
  float mean_val = 0;
  float sq_val = 0;
  int row_cnt = 0;
  int row_index = 0;
  float tmp = 0;
  float xvalue = 0;

  for (int i=0; i<FLAGS_num_cols; ++i){
    mean_val = 0;
    row_cnt = 0;
    for (auto p : Xcol.col(i)){
      xvalue = p.second;
      mean_val += xvalue; 
      row_cnt += 1;
    }
    if (row_cnt>0){
      sq_val = 0; 
      mean_val = mean_val / row_cnt;
      for (auto p : Xcol.col(i)){
        row_index = p.first;
        xvalue = p.second;
	tmp = xvalue - mean_val;
	Xcol(row_index,i) = tmp;
        sq_val += tmp*tmp;
      } 
      sq_val = sqrt(sq_val);
      if (sq_val>0){
        for (auto p : Xcol.col(i)){
          row_index = p.first;
          xvalue = p.second;
	  Xcol(row_index,i) = xvalue / sq_val;
        }
      }
    }
  }
}

void standardization_Y() {
  float mean_val = 0;
  float sq_val = 0;
  int row_cnt = 0;
  int row_index = 0;
  float tmp = 0;
  float xvalue = 0;
  for (auto p : Ycol.col(0)){
    xvalue = p.second;
    mean_val += xvalue; 
    row_cnt += 1;
  }

  if (row_cnt>0){
    sq_val = 0; 
    mean_val = mean_val / row_cnt;
    for (auto p : Ycol.col(0)){
      row_index = p.first;
      xvalue = p.second;
      tmp = xvalue - mean_val;
      Ycol(row_index,0) = tmp;
      sq_val += tmp*tmp;
    } 
    sq_val = sqrt(sq_val);
    if (sq_val>0){
      for (auto p : Ycol.col(0)){
        row_index = p.first;
        xvalue = p.second;
	Ycol(row_index,0) = xvalue / sq_val;
      }
    }
  }
}

void read_sparse_matrix_X(std::string inputfile) {
  Xcol.resize(FLAGS_num_rows,FLAGS_num_cols);
  std::ifstream inputstream(inputfile.c_str());
  while(true) {
    int row, col;
    float val;
    inputstream >> row >> col >> val;
    if (!inputstream) {
      break;
    }
    Xcol(row,col) = val;
  }
  inputstream.close();
  // standardize X. It makes the mean is zero, and 
  // variance is 1 for each column
  standardization_X();
}

// Read sparse data matrix into Xcol, Ycol. Each line of the matrix
// is a whitespace-separated triple (row,col,value), where row>=0 and col>=0.
// For example:
//
// 0 0 0.5
// 1 2 1.5
// 2 1 2.5
//
// This specifies a 3x3 matrix with 3 nonzero elements: 0.5 at (0,0), 1.5 at
// (1,2) and 2.5 at (2,1).

void read_sparse_matrix_Y(std::string inputfile) {
  Ycol.resize(FLAGS_num_rows,1);
  residual.resize(FLAGS_num_rows,1);

  std::ifstream inputstream(inputfile.c_str());
  while(true) {
    int row, col;
    float val;
    inputstream >> row >> col >> val;
    if (!inputstream) {
      break;
    }
    Ycol(row,0) = val;
  }
  inputstream.close();
  // standardize Y. It makes the mean is zero, and 
  // variance is 1 for each column
  standardization_Y();
}

void lasso_element(int a, int iter, int global_worker_id,
                 petuum::Table<float>& beta_table) {
  // get beta value
  petuum::RowAccessor beta_acc;
  beta_table.Get(0, &beta_acc);
  const petuum::DenseRow<float>& beta_val = beta_acc.Get<petuum::DenseRow<float> >();

  float new_beta = 0.0;
  float xvalue = 0.0;
  int row_index = 0;

  // Compute value for updating beta(a)
  float val_before_thr = 0.0;
  for (auto p : Xcol.col(a)){
    row_index = p.first;
    xvalue = p.second;
    val_before_thr += xvalue * residual.get(row_index,0);
  }
  val_before_thr += beta_val[a]; 
  // soft thresdholing
  if (val_before_thr>0 && FLAGS_lambda<fabs(val_before_thr)){
    new_beta = val_before_thr - FLAGS_lambda;
  } else if (val_before_thr<0 && FLAGS_lambda<fabs(val_before_thr)){
    new_beta = val_before_thr + FLAGS_lambda;
  } else {
    new_beta = 0; 
  }
  // Update residual
  for (auto p : residual.col(0)){
    residual(p.first,0) = p.second + Xcol.get(p.first,a) * (beta_val[a] - new_beta);
  }
  // Commit updates to Petuum PS
  beta_table.Inc(0,a,-beta_val[a]+new_beta);
}


// Outputs beta_table
void output_to_disk(petuum::Table<float>& beta_table) {
  std::string beta_file = FLAGS_output_prefix + ".BETA";
  std::ofstream beta_stream(beta_file.c_str());
  petuum::RowAccessor beta_acc;
  beta_table.Get(0, &beta_acc);
  const petuum::DenseRow<float>& beta_val = beta_acc.Get<petuum::DenseRow<float> >();
  for (uint32_t i = 0; i < FLAGS_num_rows; ++i) {
    beta_stream << beta_val[i] << "\n";
  }
  beta_stream.close();
}
void* solve_lasso(void* args) {
  int local_thread_id = *reinterpret_cast<int*>(args);
  // Register this thread with Petuum PS
  petuum::TableGroup::RegisterThread();
  // Get tables
  petuum::Table<float> beta_table = petuum::TableGroup::GetTableOrDie<float>(0);
  // Initialize MF solver
  const int total_num_workers = get_total_num_workers();
  const int global_worker_id = get_global_worker_id(local_thread_id); // global_worker_id lies in the range [0,total_num_workers)

  // Initialize beta table and residual
  petuum::RowAccessor beta_acc;
  beta_table.Get(0, &beta_acc);
  const petuum::DenseRow<float>& beta_val = beta_acc.Get<petuum::DenseRow<float> >();
  for (int i = 0; i < FLAGS_num_cols; ++i) {
    beta_table.Inc(0,i,-beta_val[i]);
  }
  for (int i=0; i<FLAGS_num_rows; ++i){
    residual(i,0) = Ycol.get(i,0);
  } 

  // Run additional iterations to let stale values finish propagating
  for (int iter = 0; iter < FLAGS_staleness; ++iter) {
    petuum::TableGroup::Clock();
  }

  float obj = 0;
  // Run lasso solver
  for (int iter = 0; iter < FLAGS_num_iterations; ++iter) {
    boost::posix_time::ptime beginT = boost::posix_time::microsec_clock::local_time();
    if (global_worker_id == 0) {
      std::cout << "Iteration " << iter+1 << "/" << FLAGS_num_iterations << "... " << std::flush;
    }
    // Output loss function
    if (global_worker_id == 0) {
      petuum::RowAccessor beta_acc;
      beta_table.Get(0, &beta_acc);
      const petuum::DenseRow<float>& beta_val = beta_acc.Get<petuum::DenseRow<float> >();
      obj=0;
      for (auto p : residual.col(0)){
	  obj=obj+pow(p.second,2);
      }
      obj = obj/2;
      for (int col_index=0; col_index < FLAGS_num_cols; ++col_index){
	obj += FLAGS_lambda*fabs(beta_val[col_index]);
      }
      std::cout << " loss function = " << obj << "... " << std::flush;
    }
    // Divide beta coefficients elements across workers, and perform coordinate descent
    for (int a = global_worker_id; a < FLAGS_num_cols; a += total_num_workers) {
      lasso_element(a,iter,global_worker_id,beta_table);
    }
    // Advance Parameter Server iteration
    petuum::TableGroup::Clock();
    boost::posix_time::time_duration elapTime = boost::posix_time::microsec_clock::local_time() - beginT;
    if (global_worker_id == 0) {
      std::cout << "elapsed time = " << ((float) elapTime.total_milliseconds()) / 1000 << std::endl;
    }
  }
  // Run additional iterations to let stale values finish propagating
  for (int iter = 0; iter < FLAGS_staleness; ++iter) {
    petuum::TableGroup::Clock();
  }
  // Output results to disk
  if (global_worker_id == 0) {
    std::cout << "Outputting results to prefix " << FLAGS_output_prefix << " ... " << std::flush;
    output_to_disk(beta_table);
    std::cout << "done" << std::endl;
  }
  // Deregister this thread with Petuum PS
  petuum::TableGroup::DeregisterThread();
  return 0;
}


// Main function
int main(int argc, char *argv[]) {
  google::ParseCommandLineFlags(&argc, &argv, true);
  google::InitGoogleLogging(argv[0]);
  boost::posix_time::ptime beginT = boost::posix_time::microsec_clock::local_time();
  
  // Configure Petuum PS
  petuum::TableGroupConfig table_group_config;
  // Global params
  table_group_config.num_total_server_threads = FLAGS_num_clients;  // 1 server thread per client
  table_group_config.num_total_bg_threads = FLAGS_num_clients;  // 1 background thread per client
  table_group_config.num_total_clients = FLAGS_num_clients;
  table_group_config.num_tables = 1;  // beta_table
  // Local params
  table_group_config.num_local_server_threads = 1;
  table_group_config.num_local_bg_threads = 1;
  table_group_config.num_local_app_threads = FLAGS_num_worker_threads + 1;  // +1 for main() thread
  petuum::GetHostInfos(FLAGS_hostfile, &table_group_config.host_map);
  petuum::GetServerIDsFromHostMap(&table_group_config.server_ids, table_group_config.host_map);
  table_group_config.client_id = FLAGS_client_id;
  const uint32_t threadspace_per_process = 1000;
  table_group_config.local_id_min = FLAGS_client_id * threadspace_per_process;
  table_group_config.local_id_max = (FLAGS_client_id+1) * threadspace_per_process - 1;
  table_group_config.consistency_model = petuum::SSP;
  // Configure PS row types
  petuum::TableGroup::RegisterRow<petuum::DenseRow<float> >(0);  // Register dense rows as ID 0
  // Start PS
  // IMPORTANT: This command starts up the name node service on client 0.
  //            We therefore do it ASAP, before other lengthy actions like
  //            loading data.
  petuum::TableGroup::Init(table_group_config, false);  // Initializing thread does not need table access
  
  // Load Data
  if (FLAGS_client_id == 0) {
    std::cout << "Loading data... " << std::flush;
  }

  if (FLAGS_client_id == 0) {
    std::cout << "done" << std::endl;
    // Output statistics
    std::cout << "num_rows: " << FLAGS_num_rows << std::endl;
    std::cout << "num_cols: " << FLAGS_num_cols << std::endl;
    std::cout << "input X: " << FLAGS_datafile_X << std::endl;
    std::cout << "input Y: " << FLAGS_datafile_Y << std::endl;
    std::cout << "lambda: " << FLAGS_lambda << std::endl;
    std::cout << "# number of iterations: " << FLAGS_num_iterations << std::endl;
    std::cout << "# client machines: " << FLAGS_num_clients << std::endl;
    std::cout << "# worker threads per client: " << FLAGS_num_worker_threads << std::endl;
    std::cout << "SSP staleness: " << FLAGS_staleness << std::endl;
  }
  read_sparse_matrix_X(FLAGS_datafile_X);
  read_sparse_matrix_Y(FLAGS_datafile_Y);
  
  // Configure PS tables
  petuum::ClientTableConfig table_config;
  table_config.table_info.row_type = 0; // Dense rows
  table_config.oplog_capacity = 100;
  // beta table (1 by num_cols)
  table_config.table_info.table_staleness = FLAGS_staleness;
  table_config.table_info.row_capacity = FLAGS_num_cols;
  table_config.process_cache_capacity = FLAGS_num_cols; 
  petuum::TableGroup::CreateTable(0,table_config);
  // Finished creating tables
  petuum::TableGroup::CreateTableDone();
  
  // Run Petuum PS-based Lasso solver
  pthread_t* threads = new pthread_t[FLAGS_num_worker_threads];
  int* local_thread_ids = new int[FLAGS_num_worker_threads];
  for (int t = 0; t < FLAGS_num_worker_threads; ++t) {
    local_thread_ids[t] = t;
    pthread_create(threads+t, NULL, solve_lasso, local_thread_ids+t);
  }
  petuum::TableGroup::WaitThreadRegister();
  for (int t = 0; t < FLAGS_num_worker_threads; ++t) {
    pthread_join(threads[t], NULL);
  }
  delete threads;
  delete local_thread_ids;
  
  // Cleanup and output runtime
  petuum::TableGroup::ShutDown();
  boost::posix_time::time_duration elapTime = boost::posix_time::microsec_clock::local_time() - beginT;
  if (FLAGS_client_id == 0) {
    std::cout << "total runtime = " << ((float) elapTime.total_milliseconds()) / 1000 << "s" << std::endl;
  }
  return 0;
}

File: src/main/java/com/petuum/app/Lasso_ps.java
package com.petuum.app;

import com.petuum.petuum_ps.*;
import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.OpenMapRealMatrix;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.*;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by bobo on 14-6-12.
 */
public class Lasso_ps {
    public static void main(String args[]){
        Logger log = LogManager.getLogger(Lasso_ps.class.getName());

        TableGroupConfig table_group_config = new TableGroupConfig();
        //client_table_config_map: each entry is a client table config
        Map<Integer, ClientTableConfig> client_table_config_map = new HashMap<Integer, ClientTableConfig>();
        //param_map: each entry is a parameter
        Map<String, String> param_map = new HashMap<String, String>();
        XMLparser.parseTableConfigs("config/table_Lasso.xml", table_group_config, client_table_config_map);
        XMLparser.parseParam("config/param_Lasso.xml", param_map);

        table_group_config.setClient_id(Integer.parseInt(args[0]));

        // Configure PS row types
        TableGroup.RegisterDenseFloatRow(0);         //dense row
        // Start PS
        // IMPORTANT: This command starts up the name node service on client 0.
        //            We therefore do it ASAP, before other lengthy actions like
        //            loading data.
        TableGroup.Init(table_group_config, false);
        // Load data
        // Output statistics
        int num_rows = Integer.parseInt(param_map.get("num_rows"));
        int num_cols = Integer.parseInt(param_map.get("num_cols"));
        String X_file = param_map.get("input_X");
        String Y_file = param_map.get("input_Y");
        int num_iters = Integer.parseInt(param_map.get("num_iters"));
        double lambda = Double.parseDouble(param_map.get("lambda"));
        String output_file = param_map.get("output_file");
        log.info("num_rows: " + num_rows);
        log.info("num_cols: " + num_cols);
        log.info("lambda: " + lambda);
        log.info("num_iters: " + num_iters);
        log.info("input_X: " + X_file);
        log.info("input_Y: " + Y_file);


        LassoConfig lassoConfig = new LassoConfig();
        lassoConfig.lambda = lambda;
        lassoConfig.numRows = num_rows;
        lassoConfig.numCols = num_cols;
        lassoConfig.numIters = num_iters;
        lassoConfig.outfile = output_file;

        LassoProblem lassoProblem = new LassoProblem(lassoConfig, table_group_config, client_table_config_map.get(0));
        lassoProblem.readData(X_file, Y_file);
        lassoProblem.standardiseData();

        // Configure PS tables
        Iterator iter=client_table_config_map.entrySet().iterator();
        while(iter.hasNext()){
            Map.Entry<Integer, ClientTableConfig> entry = (Map.Entry<Integer, ClientTableConfig>) iter.next();
            TableGroup.CreateTable(entry.getKey(), entry.getValue());
        }
        // Finished creating tables
        TableGroup.CreateTableDone();
        log.info("Table create done!");
        //Thread operations
        ExecutorService threadPool =
                Executors.newFixedThreadPool(table_group_config.getNum_local_app_threads() - 1);
        for(int i=0; i<table_group_config.getNum_local_app_threads()-1; i++){
            threadPool.execute(new solveLasso(i, lassoProblem));
        }
        threadPool.shutdown();
        while (!threadPool.isTerminated()) {}

//        lassoProblem.outputResult(output_file);
        log.info("Lasso_ps finished!");
    }
}
class LassoConfig{
    int numRows;
    int numCols;
    int numIters;
    double lambda;
    String outfile;
}
/**
 * Lasso regression - l1-norm
 */
class LassoProblem {
    RealMatrix Xcols;            //Xcol: store the matrix X
    ArrayRealVector Ycol;       //Ycol: store the vector Y
    ArrayRealVector Residual;   //Residual: store the residual
    LassoConfig lassoConfig;
    TableGroupConfig tableGroupConfig;
    ClientTableConfig clientTableConfig;

    public LassoProblem(LassoConfig lassoConfig,
                        TableGroupConfig tableGroupConfig,
                        ClientTableConfig clientTableConfig) {
        Xcols = new OpenMapRealMatrix(lassoConfig.numRows, lassoConfig.numCols);
        Ycol = new ArrayRealVector(lassoConfig.numRows);
        Residual = new ArrayRealVector(lassoConfig.numCols);
        this.lassoConfig = lassoConfig;

        this.tableGroupConfig = tableGroupConfig;
        this.clientTableConfig = clientTableConfig;
    }

    private static Logger log = LogManager.getLogger(Lasso_ps.class.getName());

    public void readData(String x_file, String y_file) {
        try {
            BufferedReader br_x = new BufferedReader(new FileReader(x_file));
            BufferedReader br_y = new BufferedReader(new FileReader(y_file));
            String strLine;
            while ((strLine = br_x.readLine()) != null) {
                String numArr[] = strLine.split(" ");
                Xcols.setEntry(Integer.parseInt(numArr[0]),
                        Integer.parseInt(numArr[1]),
                        Double.parseDouble(numArr[2]));
            }
            br_x.close();
            while ((strLine = br_y.readLine()) != null) {
                String numArr[] = strLine.split(" ");
                Ycol.setEntry(Integer.parseInt(numArr[0]),
                        Double.parseDouble(numArr[2]));
            }
            br_y.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void standardiseData() {
        for(int i = 0; i < Xcols.getColumnDimension(); i++){
            double mean = 0.0;
            long nonzero = 0;
            RealVector x_col = Xcols.getColumnVector(i);
            for (int j = 0; j < x_col.getDimension(); j++) {
                if (x_col.getEntry(j)!=0.0){
                    mean += x_col.getEntry(j);
                    nonzero += 1;
                }
            }
            if (nonzero != 0) {
                for (int j = 0; j < x_col.getDimension(); j++) {
                    if (x_col.getEntry(j)!=0.0)
                        x_col.addToEntry(j, -mean/nonzero);
                }

                double sqrt = x_col.getNorm();
                for (int j = 0; j < x_col.getDimension(); j++) {
                    x_col.setEntry(j, x_col.getEntry(j)/sqrt);
                }
            }
            Xcols.setColumnVector(i, x_col);
        }

        double mean_y = 0.0;
        long nonzero_y = 0;
        for(int i = 0; i < Ycol.getDimension(); i++){
            if (Ycol.getEntry(i)!=0.0) {
                mean_y += Ycol.getEntry(i);
                nonzero_y += 1;
            }
        }
        if (nonzero_y != 0){
            for (int i = 0; i < Ycol.getDimension(); i++) {
                if (Ycol.getEntry(i) != 0.0){
                    Ycol.addToEntry(i, -mean_y/nonzero_y);
                }
            }
            double sqrt = Ycol.getNorm();
            for (int i = 0; i < Ycol.getDimension(); i++) {
                Ycol.setEntry(i, Ycol.getEntry(i)/sqrt);
            }
        }
    }


}
class solveLasso implements Runnable{
    private int localThreadId;
    private LassoProblem lassoProblem;
    private static Logger log = LogManager.getLogger(Lasso_ps.class.getName());
    public solveLasso(int localThreadId, LassoProblem lassoProblem){
        this.localThreadId = localThreadId;this.lassoProblem = lassoProblem;
    }
    @Override
    public void run() {
        TableGroup.RegisterThread();
        TableFloat betaTable = TableGroup.GetTableOrDieFloat(0);
        int totalNumWorkers = lassoProblem.tableGroupConfig.getNum_total_clients()
                * (lassoProblem.tableGroupConfig.getNum_local_app_threads()-1);
        int globalWorkerId = lassoProblem.tableGroupConfig.getClient_id()
                * (lassoProblem.tableGroupConfig.getNum_local_app_threads()-1)
                + localThreadId;
        RowAccessor betaAcc = new RowAccessor();
        betaTable.Get(0, betaAcc);
        DenseFloatRow betaVal = betaAcc.GetDenseFloatRow();
        // Run additional iterations to let stale values finish propagating
        for (int i = 0; i < lassoProblem.Xcols.getColumnDimension(); i++) {
            betaTable.Inc(0, i, - betaVal.get(i));
        }
        lassoProblem.Residual = lassoProblem.Ycol.copy();
        for (int i = 0;
             i < lassoProblem.clientTableConfig.getTable_info().getTable_staleness();
             i++) {
            TableGroup.Clock();
        }
        log.info("Start lasso");
        // Run lasso solver
        for (int iter = 0; iter < lassoProblem.lassoConfig.numIters; iter++){
            long time_begin = System.currentTimeMillis();
            // Output loss function
            if (globalWorkerId == 0){
                log.info("Iteration "+(iter+1)+" ...");
                double obj = 0;
                betaTable.Get(0, betaAcc);
                betaVal = betaAcc.GetDenseFloatRow();
                obj = Math.pow(lassoProblem.Residual.getNorm(),2);
                obj /= 2;
                for (int col_i = 0; col_i < lassoProblem.lassoConfig.numCols; col_i++) {
                    obj += lassoProblem.lassoConfig.lambda * Math.abs(betaVal.get(col_i));
                }
                log.info("loss function = "+obj+" ...");
            }
            // Divide beta coefficients elements across workers, and perform coordinate descent
            for (int a = globalWorkerId; a < lassoProblem.lassoConfig.numCols;
                 a += totalNumWorkers){
                lassoElement(a, iter, betaTable);
            }
            // Advance Parameter Server iteration
            TableGroup.Clock();
            long time_end = System.currentTimeMillis();
            if (globalWorkerId == 0){
                log.info("elapsed time = "+(time_end-time_begin)/1000.0);
            }
        }
        // Run additional iterations to let stale values finish propagating
        for (int i = 0;
             i < lassoProblem.clientTableConfig.getTable_info().getTable_staleness(); i++) {
            TableGroup.Clock();
        }
        // Output data
        if(globalWorkerId == 0){
            outputResult(lassoProblem.lassoConfig.outfile);
        }
        // Deregister this thread with Petuum PS
        TableGroup.DeregisterThread();
    }

    private void lassoElement(int a, int iter, TableFloat betaTable) {
        // get beta value
        RowAccessor beta_acc = new RowAccessor();
        betaTable.Get(0, beta_acc);
        DenseFloatRow beta_val = beta_acc.GetDenseFloatRow();

        float new_beta = 0.0f;

        // Compute value for updating beta(a)
        float val_before_thr = 0.0f;
        RealVector Xcol = lassoProblem.Xcols.getColumnVector(a);
        val_before_thr = (float) Xcol.dotProduct(lassoProblem.Residual);
        val_before_thr += beta_val.get(a);
        // soft thresdholing
        if(val_before_thr>0 && lassoProblem.lassoConfig.lambda<Math.abs(val_before_thr)){
            new_beta = (float) (val_before_thr - lassoProblem.lassoConfig.lambda);
        }else if (val_before_thr<0 && lassoProblem.lassoConfig.lambda<Math.abs(val_before_thr)){
            new_beta = (float) (val_before_thr + lassoProblem.lassoConfig.lambda);
        }else
            new_beta = 0;
        // Update residual
        lassoProblem.Residual.combineToSelf(1,beta_val.get(a) - new_beta, Xcol);
        // Commit updates to Petuum PS
        betaTable.Inc(0, a, new_beta - beta_val.get(a));
    }

    public void outputResult(String output_file) {
        TableFloat betaTable = TableGroup.GetTableOrDieFloat(0);
        RowAccessor betaAcc = new RowAccessor();
        betaTable.Get(0, betaAcc);
        DenseFloatRow betaVal = betaAcc.GetDenseFloatRow();
        try {
            File out_file = new File(output_file);
            if(!out_file.getParentFile().exists())
                out_file.getParentFile().mkdirs();
            FileWriter fw = new FileWriter(output_file);
            for (int i = 0; i < lassoProblem.lassoConfig.numCols; i++) {
                fw.write(betaVal.get(i) + "\n");
            }
            fw.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
